File: sdks/python/arc_sdk/channel.py
"""
Off-chain bilateral payment channel for ARC Chain.

Provides a Channel class with pay(), receive(), close(), and dispute()
methods for high-throughput off-chain transactions between two agents.

Usage:
    channel = Channel(channel_id, "opener", total_deposit=1_000_000)
    channel.confirm_open()
    commitment = channel.pay(100)
    # Send commitment to counterparty...
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import List, Optional


class ChannelState(Enum):
    """Channel lifecycle states."""

    OPENING = "Opening"
    OPEN = "Open"
    CLOSING = "Closing"
    DISPUTED = "Disputed"
    CLOSED = "Closed"


class Role(Enum):
    """Role of a party in the channel."""

    OPENER = "opener"
    COUNTERPARTY = "counterparty"


@dataclass
class StateCommitment:
    """A signed state commitment for the channel."""

    channel_id: bytes
    nonce: int
    opener_balance: int
    counterparty_balance: int
    proposer_sig: bytes = b""
    acceptor_sig: Optional[bytes] = None

    @property
    def is_fully_signed(self) -> bool:
        return self.acceptor_sig is not None


class ChannelError(Exception):
    """Base error for channel operations."""

    pass


class Channel:
    """
    Off-chain bilateral payment channel.

    Manages state transitions, signature verification, and balance conservation
    for a two-party payment channel on ARC Chain.
    """

    def __init__(
        self,
        channel_id: bytes,
        role: Role,
        total_deposit: int,
    ):
        self.channel_id = channel_id
        self.role = role
        self.total_deposit = total_deposit
        self.state = ChannelState.OPENING

        if role == Role.OPENER:
            self.opener_balance = total_deposit
            self.counterparty_balance = 0
        else:
            self.opener_balance = 0
            self.counterparty_balance = total_deposit

        self.nonce = 0
        self._history: List[StateCommitment] = []

    def confirm_open(self) -> None:
        """Mark channel as open after on-chain ChannelOpen confirms."""
        if self.state != ChannelState.OPENING:
            raise ChannelError(f"Cannot open: channel is {self.state.value}")
        self.state = ChannelState.OPEN

    @property
    def my_balance(self) -> int:
        """Get this party's current balance."""
        if self.role == Role.OPENER:
            return self.opener_balance
        return self.counterparty_balance

    @property
    def their_balance(self) -> int:
        """Get counterparty's current balance."""
        if self.role == Role.OPENER:
            return self.counterparty_balance
        return self.opener_balance

    def pay(self, amount: int) -> StateCommitment:
        """
        Transfer `amount` from this party to the counterparty.
        Returns a half-signed state commitment to send to the counterparty.
        """
        if self.state != ChannelState.OPEN:
            raise ChannelError(f"Cannot pay: channel is {self.state.value}")

        if amount > self.my_balance:
            raise ChannelError(
                f"Insufficient balance: have {self.my_balance}, need {amount}"
            )

        if self.role == Role.OPENER:
            new_opener = self.opener_balance - amount
            new_counter = self.counterparty_balance + amount
        else:
            new_opener = self.opener_balance + amount
            new_counter = self.counterparty_balance - amount

        return self.propose_state(new_opener, new_counter)

    def propose_state(
        self, opener_balance: int, counterparty_balance: int
    ) -> StateCommitment:
        """Propose a new state with arbitrary balances."""
        if self.state != ChannelState.OPEN:
            raise ChannelError(f"Cannot propose: channel is {self.state.value}")

        if opener_balance + counterparty_balance != self.total_deposit:
            raise ChannelError(
                f"Conservation violated: {opener_balance} + {counterparty_balance} "
                f"!= {self.total_deposit}"
            )

        new_nonce = self.nonce + 1

        return StateCommitment(
            channel_id=self.channel_id,
            nonce=new_nonce,
            opener_balance=opener_balance,
            counterparty_balance=counterparty_balance,
        )

    def receive_state(self, commitment: StateCommitment) -> StateCommitment:
        """
        Receive and validate a state commitment from the counterparty.
        Updates local state if valid.
        """
        if self.state != ChannelState.OPEN:
            raise ChannelError(f"Cannot receive: channel is {self.state.value}")

        if commitment.channel_id != self.channel_id:
            raise ChannelError("Channel ID mismatch")

        if commitment.nonce <= self.nonce:
            raise ChannelError(
                f"Nonce must increase: got {commitment.nonce}, current {self.nonce}"
            )

        total = commitment.opener_balance + commitment.counterparty_balance
        if total != self.total_deposit:
            raise ChannelError("Conservation violated")

        # In full implementation, verify proposer's Ed25519 signature here.
        signed = StateCommitment(
            channel_id=commitment.channel_id,
            nonce=commitment.nonce,
            opener_balance=commitment.opener_balance,
            counterparty_balance=commitment.counterparty_balance,
            proposer_sig=commitment.proposer_sig,
            acceptor_sig=b"\x00" * 64,  # Caller signs externally
        )

        self.nonce = commitment.nonce
        self.opener_balance = commitment.opener_balance
        self.counterparty_balance = commitment.counterparty_balance
        self._history.append(signed)

        return signed

    def finalize_state(self, commitment: StateCommitment) -> None:
        """Finalize a state after receiving counterparty's co-signature."""
        if not commitment.is_fully_signed:
            raise ChannelError("Commitment not fully signed")

        self.nonce = commitment.nonce
        self.opener_balance = commitment.opener_balance
        self.counterparty_balance = commitment.counterparty_balance
        self._history.append(commitment)

    def close(self) -> StateCommitment:
        """Initiate cooperative close."""
        if self.state != ChannelState.OPEN:
            raise ChannelError(f"Cannot close: channel is {self.state.value}")

        self.state = ChannelState.CLOSING

        return StateCommitment(
            channel_id=self.channel_id,
            nonce=self.nonce,
            opener_balance=self.opener_balance,
            counterparty_balance=self.counterparty_balance,
        )

    def dispute(self) -> StateCommitment:
        """Get the latest fully-signed state for on-chain dispute submission."""
        for c in reversed(self._history):
            if c.is_fully_signed:
                return c
        raise ChannelError("No fully-signed states available for dispute")

    def confirm_closed(self) -> None:
        """Mark channel as closed after on-chain resolution."""
        self.state = ChannelState.CLOSED

    @staticmethod
    def build_open_body(
        channel_id: str,
        counterparty: str,
        deposit: int,
        timeout_blocks: int = 100,
    ) -> dict:
        """Build a ChannelOpen transaction body."""
        return {
            "type": "ChannelOpen",
            "channel_id": channel_id,
            "counterparty": counterparty,
            "deposit": deposit,
            "timeout_blocks": timeout_blocks,
        }

    @staticmethod
    def build_close_body(
        channel_id: str,
        opener_balance: int,
        counterparty_balance: int,
        counterparty_sig: list,
        state_nonce: int,
    ) -> dict:
        """Build a ChannelClose transaction body."""
        return {
            "type": "ChannelClose",
            "channel_id": channel_id,
            "opener_balance": opener_balance,
            "counterparty_balance": counterparty_balance,
            "counterparty_sig": counterparty_sig,
            "state_nonce": state_nonce,
        }

    @staticmethod
    def build_dispute_body(
        channel_id: str,
        opener_balance: int,
        counterparty_balance: int,
        other_party_sig: list,
        state_nonce: int,
        challenge_period: int = 100,
    ) -> dict:
        """Build a ChannelDispute transaction body."""
        return {
            "type": "ChannelDispute",
            "channel_id": channel_id,
            "opener_balance": opener_balance,
            "counterparty_balance": counterparty_balance,
            "other_party_sig": other_party_sig,
            "state_nonce": state_nonce,
            "challenge_period": challenge_period,
        }

File: sdks/typescript/src/channel.ts
/**
 * Off-chain bilateral payment channel for ARC Chain.
 *
 * Provides a Channel class with pay(), receive(), close(), and dispute()
 * methods for high-throughput off-chain transactions between two agents.
 *
 * @example
 * ```typescript
 * const channel = new Channel(channelId, myKeypair, counterpartyPubkey, deposit);
 * await channel.open(client);
 * const commitment = channel.pay(100);
 * // Send commitment to counterparty...
 * ```
 */

import { Hash256, Address, TransactionBody } from './types';

/** Channel lifecycle states. */
export enum ChannelState {
  Opening = 'Opening',
  Open = 'Open',
  Closing = 'Closing',
  Disputed = 'Disputed',
  Closed = 'Closed',
}

/** A signed state commitment for the channel. */
export interface StateCommitment {
  channelId: Hash256;
  nonce: number;
  openerBalance: number;
  counterpartyBalance: number;
  /** Signature from the party that proposed this state. */
  proposerSig: Uint8Array;
  /** Signature from the party that accepted this state (null if pending). */
  acceptorSig: Uint8Array | null;
}

/** Role of a party in the channel. */
export type Role = 'opener' | 'counterparty';

/**
 * Off-chain bilateral payment channel.
 *
 * Manages state transitions, signature verification, and balance conservation
 * for a two-party payment channel on ARC Chain.
 */
export class Channel {
  readonly channelId: Hash256;
  readonly role: Role;
  readonly totalDeposit: number;

  state: ChannelState = ChannelState.Opening;
  openerBalance: number;
  counterpartyBalance: number;
  nonce: number = 0;

  private history: StateCommitment[] = [];

  constructor(
    channelId: Hash256,
    role: Role,
    totalDeposit: number,
  ) {
    this.channelId = channelId;
    this.role = role;
    this.totalDeposit = totalDeposit;

    if (role === 'opener') {
      this.openerBalance = totalDeposit;
      this.counterpartyBalance = 0;
    } else {
      this.openerBalance = 0;
      this.counterpartyBalance = totalDeposit;
    }
  }

  /** Mark channel as open after on-chain ChannelOpen confirms. */
  confirmOpen(): void {
    if (this.state !== ChannelState.Opening) {
      throw new Error(`Cannot open: channel is ${this.state}`);
    }
    this.state = ChannelState.Open;
  }

  /** Get this party's current balance. */
  myBalance(): number {
    return this.role === 'opener' ? this.openerBalance : this.counterpartyBalance;
  }

  /** Get counterparty's current balance. */
  theirBalance(): number {
    return this.role === 'opener' ? this.counterpartyBalance : this.openerBalance;
  }

  /**
   * Transfer `amount` from this party to the counterparty.
   * Returns a half-signed state commitment to send to the counterparty.
   */
  pay(amount: number): StateCommitment {
    if (this.state !== ChannelState.Open) {
      throw new Error(`Cannot pay: channel is ${this.state}`);
    }

    if (amount > this.myBalance()) {
      throw new Error(`Insufficient balance: have ${this.myBalance()}, need ${amount}`);
    }

    let newOpener: number, newCounter: number;
    if (this.role === 'opener') {
      newOpener = this.openerBalance - amount;
      newCounter = this.counterpartyBalance + amount;
    } else {
      newOpener = this.openerBalance + amount;
      newCounter = this.counterpartyBalance - amount;
    }

    return this.proposeState(newOpener, newCounter);
  }

  /** Propose a new state with arbitrary balances. */
  proposeState(openerBalance: number, counterpartyBalance: number): StateCommitment {
    if (this.state !== ChannelState.Open) {
      throw new Error(`Cannot propose: channel is ${this.state}`);
    }

    if (openerBalance + counterpartyBalance !== this.totalDeposit) {
      throw new Error(
        `Conservation violated: ${openerBalance} + ${counterpartyBalance} != ${this.totalDeposit}`
      );
    }

    const newNonce = this.nonce + 1;

    const commitment: StateCommitment = {
      channelId: this.channelId,
      nonce: newNonce,
      openerBalance,
      counterpartyBalance,
      proposerSig: new Uint8Array(64), // Caller signs externally
      acceptorSig: null,
    };

    return commitment;
  }

  /**
   * Receive and validate a state commitment from the counterparty.
   * Updates local state if valid.
   */
  receiveState(commitment: StateCommitment): StateCommitment {
    if (this.state !== ChannelState.Open) {
      throw new Error(`Cannot receive: channel is ${this.state}`);
    }

    if (commitment.channelId !== this.channelId) {
      throw new Error('Channel ID mismatch');
    }

    if (commitment.nonce <= this.nonce) {
      throw new Error(`Nonce must increase: got ${commitment.nonce}, current ${this.nonce}`);
    }

    if (commitment.openerBalance + commitment.counterpartyBalance !== this.totalDeposit) {
      throw new Error('Conservation violated');
    }

    // In a full implementation, verify proposer's Ed25519 signature here.
    // For now, accept the commitment and co-sign.

    const signed: StateCommitment = {
      ...commitment,
      acceptorSig: new Uint8Array(64), // Caller signs externally
    };

    this.nonce = commitment.nonce;
    this.openerBalance = commitment.openerBalance;
    this.counterpartyBalance = commitment.counterpartyBalance;
    this.history.push(signed);

    return signed;
  }

  /** Finalize a state after receiving counterparty's co-signature. */
  finalizeState(commitment: StateCommitment): void {
    if (!commitment.acceptorSig) {
      throw new Error('Commitment not fully signed');
    }

    this.nonce = commitment.nonce;
    this.openerBalance = commitment.openerBalance;
    this.counterpartyBalance = commitment.counterpartyBalance;
    this.history.push(commitment);
  }

  /** Initiate cooperative close. */
  close(): StateCommitment {
    if (this.state !== ChannelState.Open) {
      throw new Error(`Cannot close: channel is ${this.state}`);
    }

    this.state = ChannelState.Closing;

    return {
      channelId: this.channelId,
      nonce: this.nonce,
      openerBalance: this.openerBalance,
      counterpartyBalance: this.counterpartyBalance,
      proposerSig: new Uint8Array(64),
      acceptorSig: null,
    };
  }

  /** Get the latest fully-signed state for on-chain dispute submission. */
  dispute(): StateCommitment {
    const latest = [...this.history].reverse().find((c) => c.acceptorSig !== null);
    if (!latest) {
      throw new Error('No fully-signed states available for dispute');
    }
    return latest;
  }

  /** Mark channel as closed after on-chain resolution. */
  confirmClosed(): void {
    this.state = ChannelState.Closed;
  }

  /** Build a ChannelOpen transaction body. */
  static buildOpenBody(
    channelId: Hash256,
    counterparty: Address,
    deposit: number,
    timeoutBlocks: number = 100,
  ): TransactionBody {
    return {
      type: 'ChannelOpen',
      channel_id: channelId,
      counterparty,
      deposit,
      timeout_blocks: timeoutBlocks,
    };
  }

  /** Build a ChannelClose transaction body. */
  static buildCloseBody(
    channelId: Hash256,
    openerBalance: number,
    counterpartyBalance: number,
    counterpartySig: number[],
    stateNonce: number,
  ): TransactionBody {
    return {
      type: 'ChannelClose',
      channel_id: channelId,
      opener_balance: openerBalance,
      counterparty_balance: counterpartyBalance,
      counterparty_sig: counterpartySig,
      state_nonce: stateNonce,
    };
  }

  /** Build a ChannelDispute transaction body. */
  static buildDisputeBody(
    channelId: Hash256,
    openerBalance: number,
    counterpartyBalance: number,
    otherPartySig: number[],
    stateNonce: number,
    challengePeriod: number = 100,
  ): TransactionBody {
    return {
      type: 'ChannelDispute',
      channel_id: channelId,
      opener_balance: openerBalance,
      counterparty_balance: counterpartyBalance,
      other_party_sig: otherPartySig,
      state_nonce: stateNonce,
      challenge_period: challengePeriod,
    };
  }
}
